File: Dart/15-Greedy/Problems/1-JumpGame/JumpGame.dart
import 'dart:math';

void main() {
  // Example input arrays
  List<int> nums1 = [2, 3, 1, 1, 4];
  List<int> nums2 = [3, 2, 1, 0, 4];
  List<int> nums3 = [0];
  List<int> nums4 = [2, 0, 0];
  List<int> nums5 = [2, 5, 0, 0];

  // --- Approach 1: Recursive (Top-Down) ---
  print("\nApproach 1: Recursive (Top-Down)");
  print("Can jump to the end for ${nums1}: ${canJumpRecursive(nums1, 0)}"); // true
  print("Can jump to the end for ${nums2}: ${canJumpRecursive(nums2, 0)}"); // false
  print("Can jump to the end for ${nums3}: ${canJumpRecursive(nums3, 0)}"); // true
  print("Can jump to the end for ${nums4}: ${canJumpRecursive(nums4, 0)}"); // false
  print("Can jump to the end for ${nums5}: ${canJumpRecursive(nums5, 0)}"); // true

  // --- Approach 2: Recursive with Memoization (Top-Down) ---
  print("\nApproach 2: Recursive with Memoization (Top-Down)");
  print("Can jump to the end for ${nums1}: ${canJumpMemoized(nums1)}"); // true
  print("Can jump to the end for ${nums2}: ${canJumpMemoized(nums2)}"); // false
  print("Can jump to the end for ${nums3}: ${canJumpMemoized(nums3)}"); // true
  print("Can jump to the end for ${nums4}: ${canJumpMemoized(nums4)}"); // false
  print("Can jump to the end for ${nums5}: ${canJumpMemoized(nums5)}"); // true

  // --- Approach 3: Iterative (Bottom-Up) - Dynamic Programming ---
  print("\nApproach 3: Iterative (Bottom-Up) - Dynamic Programming");
  print("Can jump to the end for ${nums1}: ${canJumpIterativeDP(nums1)}"); // true
  print("Can jump to the end for ${nums2}: ${canJumpIterativeDP(nums2)}"); // false
  print("Can jump to the end for ${nums3}: ${canJumpIterativeDP(nums3)}"); // true
  print("Can jump to the end for ${nums4}: ${canJumpIterativeDP(nums4)}"); // false
  print("Can jump to the end for ${nums5}: ${canJumpIterativeDP(nums5)}"); // true

  // --- Approach 4: Greedy ---
  print("\nApproach 4: Greedy");
  print("Can jump to the end for ${nums1}: ${canJumpGreedy(nums1)}"); // true
  print("Can jump to the end for ${nums2}: ${canJumpGreedy(nums2)}"); // false
  print("Can jump to the end for ${nums3}: ${canJumpGreedy(nums3)}"); // true
  print("Can jump to the end for ${nums4}: ${canJumpGreedy(nums4)}"); // false
  print("Can jump to the end for ${nums5}: ${canJumpGreedy(nums5)}"); // true

  // --- Approach 5: Optimized Greedy ---
  print("\nApproach 5: Optimized Greedy");
  print("Can jump to the end for ${nums1}: ${canJumpGreedyOptimized(nums1)}");
  print("Can jump to the end for ${nums2}: ${canJumpGreedyOptimized(nums2)}");
  print("Can jump to the end for ${nums3}: ${canJumpGreedyOptimized(nums3)}");
  print("Can jump to the end for ${nums4}: ${canJumpGreedyOptimized(nums4)}");
  print("Can jump to the end for ${nums5}: ${canJumpGreedyOptimized(nums5)}");
}

// Approach 1: Recursive (Top-Down)
// Time Complexity: O(2^n) - Exponential
// Space Complexity: O(n) - Call stack depth
bool canJumpRecursive(List<int> nums, int position) {
  if (position == nums.length - 1) {
    return true; // Reached the end, we can jump!
  }

  int maxJump = nums[position]; // Max distance we can jump from current position
  for (int i = 1; i <= maxJump; i++) {
    // Try all possible jumps from the current position
    int nextPosition = position + i;
    if (nextPosition < nums.length) {
      if (canJumpRecursive(nums, nextPosition)) {
        return true; // If any jump leads to the end, we're good
      }
    }
  }
  return false; // No jump leads to the end
}

// Approach 2: Recursive with Memoization (Top-Down)
// Time Complexity: O(n^2)
// Space Complexity: O(n) - Memoization array and call stack
bool canJumpMemoized(List<int> nums) {
  List<bool?> memo = List.filled(nums.length, null); // Use null to represent unvisited
  return _canJumpMemoizedHelper(nums, 0, memo);
}

bool _canJumpMemoizedHelper(List<int> nums, int position, List<bool?> memo) {
  if (position == nums.length - 1) {
    return true;
  }

  if (memo[position] != null) {
    return memo[position]!; // Return the memoized result
  }

  int maxJump = nums[position];
  for (int i = 1; i <= maxJump; i++) {
    int nextPosition = position + i;
    if (nextPosition < nums.length) {
      if (_canJumpMemoizedHelper(nums, nextPosition, memo)) {
        memo[position] = true; // Memoize the result
        return true;
      }
    }
  }
  memo[position] = false; // Memoize the result
  return false;
}

// Approach 3: Iterative (Bottom-Up) - Dynamic Programming
// Time Complexity: O(n^2)
// Space Complexity: O(n)
bool canJumpIterativeDP(List<int> nums) {
  List<bool> dp = List.filled(nums.length, false);
  dp[nums.length - 1] = true; // Last position is reachable

  for (int i = nums.length - 2; i >= 0; i--) {
    int maxJump = nums[i];
    for (int j = 1; j <= maxJump; j++) {
      int nextPosition = i + j;
      if (nextPosition < nums.length && dp[nextPosition]) {
        dp[i] = true; // If any next position is reachable, current is also reachable
        break; // No need to check further jumps
      }
    }
  }
  return dp[0]; // Is the first position reachable?
}

// Approach 4: Greedy
// Time Complexity: O(n)
// Space Complexity: O(1)
bool canJumpGreedy(List<int> nums) {
  int maxReach = 0; // Furthest position we can reach
  for (int i = 0; i < nums.length; i++) {
    if (i > maxReach) {
      return false; // If current position is beyond maxReach, we can't reach here
    }
    maxReach = max(maxReach, i + nums[i]); // Update maxReach
  }
  return true; // If we get through the loop, we can reach the end
}

// Approach 5: Optimized Greedy
// Time Complexity: O(n)
// Space Complexity: O(1)
bool canJumpGreedyOptimized(List<int> nums) {
  int goal = nums.length - 1; // Start from the goal (last index)

  for (int i = nums.length - 2; i >= 0; i--) {
    if (i + nums[i] >= goal) {
      goal = i; // Update the goal to the current position
    }
  }
  return goal == 0; // If the goal is the first position, we can reach the end
}

File: Java/15-Greedy/Problems/1-JumpGame/JumpGame.java
import java.util.Arrays;

public class JumpGame {

    /**
     * Approach 1: Greedy Approach (Most Efficient)
     *
     * Time Complexity: O(n), where n is the length of the array.
     * Space Complexity: O(1).
     *
     * Explanation:
     * The greedy approach works by iterating through the array and keeping track of the
     * farthest reachable index. At each step, we update the farthest reachable index
     * and check if we can reach the end of the array.
     */
    public boolean canJumpGreedy(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false; // Handle null or empty input
        }
        int maxReach = 0; // The farthest index that can be reached
        for (int i = 0; i < nums.length; i++) {
            if (i > maxReach) {
                return false; // If the current index is beyond the maximum reach, we can't reach it.
            }
            maxReach = Math.max(maxReach, i + nums[i]); // Update the maximum reach
            if (maxReach >= nums.length - 1) {
                return true; // If the maximum reach is greater than or equal to the last index, we can reach the end.
            }
        }
        return false; // If the loop finishes without reaching the end, we can't jump to the end.
    }

    /**
     * Approach 2: Dynamic Programming (Top-Down with Memoization)
     *
     * Time Complexity: O(n), where n is the length of the array.
     * Space Complexity: O(n) for the memoization array.
     *
     * Explanation:
     * This approach uses recursion with memoization to avoid redundant calculations.
     * We start from the beginning of the array and recursively check if we can reach
     * the end. The memoization array stores the result for each index to avoid
     * recomputing it.
     */
    public boolean canJumpDPTopDown(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        int[] memo = new int[nums.length];
        Arrays.fill(memo, -1); // -1: unknown, 0: false, 1: true
        memo[nums.length - 1] = 1; // The last index is reachable

        return canJumpFrom(0, nums, memo);
    }

    private boolean canJumpFrom(int position, int[] nums, int[] memo) {
        if (memo[position] != -1) {
            return memo[position] == 1; // Return the memoized result
        }

        int maxJump = Math.min(position + nums[position], nums.length - 1); // Max distance we can jump
        for (int i = maxJump; i > position; i--) {
            if (canJumpFrom(i, nums, memo)) {
                memo[position] = 1; // Mark the current position as reachable
                return true;
            }
        }

        memo[position] = 0; // Mark the current position as unreachable
        return false;
    }

    /**
     * Approach 3: Dynamic Programming (Bottom-Up)
     *
     * Time Complexity: O(n^2), where n is the length of the array.
     * Space Complexity: O(n) for the dp array.
     *
     * Explanation:
     * This approach uses a bottom-up dynamic programming approach. We create a dp array
     * where dp[i] indicates whether we can reach index i. We start from the end of the
     * array and work our way back to the beginning.
     * This approach is less efficient than the greedy approach due to the nested loop.
     */
    public boolean canJumpDPBottomUp(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        boolean[] dp = new boolean[nums.length];
        dp[nums.length - 1] = true; // The last index is reachable

        for (int i = nums.length - 2; i >= 0; i--) {
            int maxJump = Math.min(i + nums[i], nums.length - 1);
            for (int j = i + 1; j <= maxJump; j++) {
                if (dp[j]) {
                    dp[i] = true; // If we can reach any of the next reachable indices, we can reach the current index.
                    break;
                }
            }
        }
        return dp[0]; // Check if we can reach the first index.
    }

    /**
     * Approach 4: Backtracking (Brute Force - Time Limit Exceeded for larger inputs)
     *
     * Time Complexity: O(2^n) in the worst case, where n is the length of the array.
     * Space Complexity: O(n) for the recursion stack.
     *
     * Explanation:
     * This is a brute-force approach that tries all possible paths. It recursively
     * explores all possible jumps from each position. This approach is very inefficient
     * and will result in a time limit exceeded error for larger input arrays.
     */
    public boolean canJumpBacktracking(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        return canJumpFromBacktracking(0, nums);
    }

    private boolean canJumpFromBacktracking(int position, int[] nums) {
        if (position == nums.length - 1) {
            return true; // Base case: If we reach the last index, we are done.
        }

        int maxJump = Math.min(position + nums[position], nums.length - 1);
        for (int i = maxJump; i > position; i--) {
            if (canJumpFromBacktracking(i, nums)) {
                return true; // If we can reach the end from any of the next positions, return true.
            }
        }
        return false; // If we can't reach the end from any of the next positions, return false.
    }

    /**
     * Approach 5: Iterative Greedy (Similar to Approach 1, but with slight variation)
     *
     * Time Complexity: O(n), where n is the length of the array.
     * Space Complexity: O(1).
     *
     * Explanation:
     * This is another greedy approach, similar to the first one, but it iterates backwards.
     * It starts from the end and checks if any previous position can reach the current position.
     */
    public boolean canJumpIterativeGreedy(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        int lastReachable = nums.length - 1; // Initialize the last reachable position to the end.
        for (int i = nums.length - 2; i >= 0; i--) {
            if (i + nums[i] >= lastReachable) {
                lastReachable = i; // Update the last reachable position if the current position can reach it.
            }
        }
        return lastReachable == 0; // If the first position can reach the end, return true.
    }

    public static void main(String[] args) {
        JumpGame solution = new JumpGame();

        // Test cases
        int[] nums1 = {2, 3, 1, 1, 4};
        System.out.println("Can jump to the end (Greedy): " + solution.canJumpGreedy(nums1)); // Output: true
        System.out.println("Can jump to the end (DP Top-Down): " + solution.canJumpDPTopDown(nums1)); // Output: true
        System.out.println("Can jump to the end (DP Bottom-Up): " + solution.canJumpDPBottomUp(nums1)); // Output: true
        System.out.println("Can jump to the end (Backtracking): " + solution.canJumpBacktracking(nums1)); // Output: true
        System.out.println("Can jump to the end (Iterative Greedy): " + solution.canJumpIterativeGreedy(nums1)); // Output: true


        int[] nums2 = {3, 2, 1, 0, 4};
        System.out.println("Can jump to the end (Greedy): " + solution.canJumpGreedy(nums2)); // Output: false
        System.out.println("Can jump to the end (DP Top-Down): " + solution.canJumpDPTopDown(nums2)); // Output: false
        System.out.println("Can jump to the end (DP Bottom-Up): " + solution.canJumpDPBottomUp(nums2)); // Output: false
        System.out.println("Can jump to the end (Backtracking): " + solution.canJumpBacktracking(nums2)); // Output: false
        System.out.println("Can jump to the end (Iterative Greedy): " + solution.canJumpIterativeGreedy(nums2)); // Output: false

        int[] nums3 = {0};
        System.out.println("Can jump to the end (Greedy): " + solution.canJumpGreedy(nums3)); // Output: true
        System.out.println("Can jump to the end (DP Top-Down): " + solution.canJumpDPTopDown(nums3)); // Output: true
        System.out.println("Can jump to the end (DP Bottom-Up): " + solution.canJumpDPBottomUp(nums3)); // Output: true
        System.out.println("Can jump to the end (Backtracking): " + solution.canJumpBacktracking(nums3)); // Output: true
        System.out.println("Can jump to the end (Iterative Greedy): " + solution.canJumpIterativeGreedy(nums3)); // Output: true

        int[] nums4 = {2,0,0};
        System.out.println("Can jump to the end (Greedy): " + solution.canJumpGreedy(nums4)); // Output: true
        System.out.println("Can jump to the end (DP Top-Down): " + solution.canJumpDPTopDown(nums4)); // Output: true
        System.out.println("Can jump to the end (DP Bottom-Up): " + solution.canJumpDPBottomUp(nums4)); // Output: true
        System.out.println("Can jump to the end (Backtracking): " + solution.canJumpBacktracking(nums4)); // Output: true
        System.out.println("Can jump to the end (Iterative Greedy): " + solution.canJumpIterativeGreedy(nums4)); // Output: true
    }
}

